### ISSUE TITLE
What occurred around Oct 06 2018 8pm server time? Hunter 3:2 now dead.

### ISSUE THREAD

**Issue by amarathTESTER** [SunwellWoW/Sunwell-TBC-Bugtracker#590]
Decription: 
As the title indicates "What occurred around Oct 06 2018 8pm server time? "

Before this 3:2 hunter rotation appeared to be working with weapons of 28 speed and higher. (Broken at 2.7 and lower; which is just fine). First off i don't care 1:1 still works. But i told people on discord i would test it and test it i did, until well...
Something occurred that changed the queuing or gcd around that time, that changed the macro from what looked to be 1 shot a second (WORKING), .75 shots a second 3:2 rotation, to steady shot spam with rare if any auto shots (100% steady shots).

How it works: 

Barebones Macro:

#showtooltip Steady Shot
/cast !Auto Shot
/cast Steady shot
After that time you only get steady shot spams (I am assuming someone applied Haste to GCD? Maybe only certain sources of Haste? Off for sure.)

How it should work: 

(-.5s GCD starts, auto fire starts [hit from previous rotation])
1.0s GCD finishes - GCD finishes before the steady shot cast time allowing the 2nd shot to be cast right away
1.1s steady [hit]
2.2s steady [hit] (the 1.5 GCD from the previous Steady Shot ENDED at 1.0s from when you sent the spell to the server from the start of previous rotations steady or autoshot).
2.6s GCD finishes, GCD starts for next steady
2.7s auto - [hit] (delayed from steady shot, its a 0.5 animation)
3.8s steady - [hit] (this is starts at 2.6s and the GCD starts then also but it takes the 0.5 animation from the auto longer to get off so ~1.15s steady shot+auto cast)
4.1s GCD finishes (This started at 2.6. 2.6+1.5 = 4.1; see what Haste (at least item and talent.. buff might, but don't think so) FOR SURE DOESN'T EFFECT GCD.)

5.2 steady    \
or                   > (This depends on the rotation you are on).
4.8 auto       /

Source (you should point out proofs of your report, please give us some source): 
Retail functioning https://web.archive.org/web/20080211131225/http://elitistjerks.com:80/f31/t11420-hunter_beast_mastery_bible/p109/

Here we can see the current state, after a whatever changed. (That one auto i let through on the end just to show the macro was asking for them.)
![32-broken](https://user-images.githubusercontent.com/43596223/46577796-bc00be00-c9bc-11e8-90c7-def7d30d4a99.jpg)

Here is how it used to look with a 2.9 speed weapon:
![3-2-b](https://user-images.githubusercontent.com/43596223/46577802-0a15c180-c9bd-11e8-97e8-c71e0c742e5f.jpg)

Even tested with a slower weapon of 3.0 (its normal for the ratio to change to more autos as speeds slow [also found in those linked EJ pages]; however the server changed before i could test if it was getting 1 shot a second - maybe 2.9 weapon just got lucky with the timing):  
![3-2-3-0](https://user-images.githubusercontent.com/43596223/46577817-80b2bf00-c9bd-11e8-81e4-d69d2ccef945.jpg)

Despite the positive outlook of 3:2, on Sunwell's TBC server, on 2.7 speed and lower weapons FOR SURE DIDN'T WORK:
![32-2-7-b](https://user-images.githubusercontent.com/43596223/46577830-c2dc0080-c9bd-11e8-84dd-d0a15f6c045e.jpg)

While this might look similar to the 3.0 speed its should have MORE steady shots not less than the 2.9 speed weapon.

I am going to put this on the beta issue forums also so if you wish to ask me questions you can there or ask Nadiir for who i am on discord. Like i stated no big to me but at least you now have this.

**** Another issue might be that once an auto shot starts it shouldn't be interruptible by casting steady shot, i believe. It appears to on beta. Maybe autos are trying to go off but are interrupted. But its really hard to tell.

**Comment by lukaasm** [comment]
Nah, that is other issue, not related to GCD. There is no grace period ( first 500ms ) during ranged spell cast in which autoshoot can still fire/not be interrupted, which makes Shoot to be always delayed due to another cast in progress

I am currently trying to fix it, but there is still one client-side issue to be resolved.

**Comment by amarathTESTER** [comment]
Tested at multiple speeds and with both 1:1 and 3:2 rotations. 3:2 appears to be at least doing a 3:2, but not what it did on retail. I think the slower speed weapons appear to be losing too many rotations but, might just be a compounding latency issue. The 3:2 rotation on retail increased Damage for all slower weapons. It largely remains the same. That being said even the 1:1 rations were not hitting this mark. Will test again. but this time with more focus on hunter dps.

**Comment by Wojtyas** [comment]
@amarathTESTER Any news?

**Comment by amarathTESTER** [comment]
TLDR: The hits per second are in-line with the 1:1 rotation, but at them moment appear off in 3:2 ill test it live or on a dev realm if you guys have one later. Maybe missing a few autos (could be ~5% less dps). But i may have introduced bias into the testing (different sources of haste). **But 3:2 is not far off retail, if it is off at all. So good job.** 

**It appears that you will be able to use your slower than 2.7 ranged weapons with the 3:2. And I can say despite any itemization errors, my hpm (hits per minute are correct). 

So first off 3:2 was never real. Haha. 3:2 is plagued with things like, if your MS is 80-100, a 2.7 speed weapon and using KC (Kill command) you will get less autos and hps (ratio of 3.4:2 to 5.6:2). Where as, a 2.8 speed weapon provides KCs with little hps and ratio change.  More on this much later. And the dps really only increases after tier set bonuses and higher scaling of steady shot over auto (mana too).

I replicated the frequency of the 1:1 rotation. That is working.  I'll include some data and info i did gather, but recognize it is potentially biased.

-------------------------------
The rest of the post is simply to preserve some of the information on what is occurring.

On sunwell beta nov 12ish

My latency is 110ms ish which is actually ideal for this as the 1.93 is a bit too fast for say 10ms per EJ.
In 1 minute: 
-3:2- (not blizlike as the autos are the things that should be changing; but 1 min test are prone to error.)
34 SS, 18 auto, 2.27 speed. 52 hpm
35 SS, 18 auto, 2.03 speed. 53 hpm
36 SS, 18 auto, 1.92 speed. 54 hpm

-1:1- (This is blizlike)
30 SS, 30 auto, 1.92 speed. 60 hpm
29 SS, 29 auto, 2.03 speed. 58 hpm
27 SS, 26 auto, 2.27 speed. 53 hpm

10 minutes:
-3:2-
360 SS, 185 auto, 2.18 speed. ~54.5 hpm
360 SS, 185 auto, 2.12 speed. 54.5 hpm

-1:1-
294 SS, 294 auto, 2.03. ~59 hpm

The rest followed this trend. 


-----The Retail MATH----- (thanks to Nosti from 08)

-Max hits per minute-
60 sec /1.5 + 0.05 (click delay) = 38.7 steady shots a minute.
60s/ weapon speed + 0.05 = 29.9 autoshots  a minute.
68.6 hpm (1.14 hps) is the MAX hps possible if no autoshots where ever interrupted (without haste and a 2.7 speed weapon), but they are QQ. So what you can see here is that while Steady Shots shouldn't increase in number AutoShots CAN with more haste. 


-What should my haste be?-
You are always better faster than slower (save clipping with KC). say you get a 50ms jump you might start clipping a ton of autoshots. 1.6s per steadyshot. Its possible that its 1.5 per steady shot, i think it is.
There are many break points like how hots at times could gain an extra tick.

Rule: (SS + autoshot) < 3.2s   (Largely a 3:2 rotation)
1.09 + 2.17 (plus 0.05 for ) =  3.31  So this will clip the crap out of autos.
But with 74+ haste get sub 3.2 which means  a 3:2 rotation.
1.04 + 2.08 (+0.05) =3.17 which is <3.2 

Next we can go up another 1.6s so 4.8s.

(SS  + AutoShot * 2) < 4.8s   (4:3 rotation)
IAotH proc:
1.09 + 3.78 (+0.1) = 4.97s (fail)               no haste
1.04 + 3.62 (+0.1) = 4.76s < 4.8s            w/ 74 haste

https://web.archive.org/web/20081009161505/http://elitistjerks.com/f31/t11420-hunter_beast_mastery_bible/p123/#post671793

If you want to see more detailed info:
https://web.archive.org/web/20080321233813/http://elitistjerks.com:80/673256-post3087.html

https://web.archive.org/web/20081009161515/http://elitistjerks.com/f31/t11420-hunter_beast_mastery_bible/p125/#post674259

---

Pretty sure this is how it works, basically the server doesn't know when you are casting SS but it does know when you can cast autshot again. So it can start casting autoshot server side and as long as it don't try to fire during KC or steady shot it will technically do the 0.5 autoshot cast while you are also casting steadyshot (so the end of the autoshot must end after your steadyshot is done).

https://web.archive.org/web/20080924141707/http://elitistjerks.com/f31/t11420-hunter_beast_mastery_bible/p121/#post661187

----

If you look into the issue people say slower weapons clip more autos and its true but they also clip less KCs. You largely get one or the other, and later with SS hitting harder than autoshots and a small loss in actual hps higher SS to auto ratios are GOOD as long as your hps doesn't suffer.

---------

1k shot table of KC and weapon speeds - differs from most others findings, but isn't too far off
https://web.archive.org/web/20081110152352/http://elitistjerks.com/f31/t11420-hunter_beast_mastery_bible/p117/
